docs/DafnyRef/UserGuide.md

```diff
@@ -156,14 +156,227 @@ exit code.
+
+* any explicit [`assert` statement](#sec-assert-statement) is _an assertion_[^precision-requires-clause].
+* A consecutive pair of lines in a [`calc` statement](#sec-calc-statement) forms _an assertion_.
+* Every function or method call with a [`requires` clause](#sec-requires-clause) yields _one assertion per requires clause_[^precision-requires-clause]
+  (special cases such as sequence indexing come with a special require clause that the index is within bounds).
+* Assignments `o.f := E;` yield an _assertion_ that `o.f` is allowed by the enclosing [`modifies` clause](#sec-loop-framing).
+* [Assign-such-that operators](#sec-update-and-call-statement) `x :| P(x)` yield an _assertion_ that `exists x :: P(x)`.
+* Every [`ensures` clause](#sec-ensures-clause) yields an _assertion_ at the end of the method and on every return.
+* Every [`decreases` clause](#sec-decreases-clause) yields an _assertion_ at either a call site or at the end of a while loop.
+* Every non-empty [subset type](#sec-subset-types) yields an _assertion_ that its witness satisfies the constraint.
+
+[^precision-requires-clause]: Dafny actually breaks things down further. For example, a precondition `requires A && B` or an assert statement `assert A && B;` turns into two assertions, more or less like `requires A requires B` and `assert A; assert B;`.
+
+It is useful to mentally visualize all these assertions as a list (an _assertion batch_) that roughly follows the order in the code[^complexity-path-encoding],
+except for `ensures` assertions that appear earlier in the code but, for verification purposes, would appear at the end.
+Thus, to prove or disprove a single assertion within this list, all the assumptions Dafny needs are 1) the global context,
+and 2) every preceding assumption (and previous assertions transformed in assumptions).
+
+[^complexity-path-encoding]: All the complexities of the execution paths (if-then-else, loops, goto, break....) are, down the road and for verification purposes, cleverly encoded with variables recording the paths and guarding assumptions made on each path. In practice, a second clever encoding of variables enables grouping many assertions together, and recovers which assertion is failing based on the value of variables that the SMT solver returns.
+
+To reduce overhead, Dafny collects all the assertions of one method into one single conjecture that it sends to the verifier, which tries to prove it correct:
+
+* If the verifier says it is correct[^smt-encoding], it means that all the assertions hold.
+* If the verifier returns a counterexample, this counter-example is used to determine both the failing assertion and the failing path.
+  In order to retrieve additional failing assertions, Dafny will again query the verifier after turning previously failed assertions into assumptions, [like this](https://github.com/dafny-lang/dafny/discussions/1898#discussioncomment-2344533).[^caveat-about-assertion-and-assumption]
```

Comment:
I wouldn't link to a discussion from here - there's no telling what content other users will add there in the future and be implicitly endorsed as correct by this link. 

In any case this feels like detail we don't need at this point. The bullet points are all implementation detail and can be moved to footnotes or elsewhere. The important point is just that an assertion batch is verified all at once and produces a single result. See my proposed edit above.

Reply:
I am linking the post, so let me 1) make it explicitly known that I link to the post and not the discussion and 2) move this into a footnote, which have less endorsement than a link.